packages/compiler/src/compiler/readMetadata.ts

```diff
@@ -118,134 +169,208 @@ export class ReadMetadata {
+        )
       }
-      if (node.key) {
-        const keyScopeContext = copyScopeContext(scopeContext)
-        keyScopeContext.definedVariables.add(contextVarName)
-        await this.updateScopeContext({ node: node.key, scopeContext })
+      if (indexVarName && scopeContext.definedVariables.has(indexVarName)) {
+        throw this.expressionError(
+          errors.variableAlreadyDeclared(indexVarName),
+          node.index!,
+        )
       }
 
+      const iterableScope = copyScopeContext(scopeContext)
+      iterableScope.definedVariables.add(contextVarName)
+      if (indexVarName) {
+        iterableScope.definedVariables.add(indexVarName)
+      }
       for await (const childNode of node.children ?? []) {
-        const childScopeContext = copyScopeContext(scopeContext)
-        childScopeContext.definedVariables.add(contextVarName)
-        if (indexVarName) childScopeContext.definedVariables.add(indexVarName)
-
         await this.readBaseMetadata({
           node: childNode,
-          scopeContext: copyScopeContext(scopeContext),
+          scopeContext: iterableScope,
           isInMessageTag,
           isInContentTag,
         })
       }
-      if (node.else) {
-        for await (const childNode of node.else.children ?? []) {
+      return
+    }
+
+    if (node.type === 'ElementTag') {
+      if (node.name === TOOL_CALL_TAG) {
+        if (isInContentTag) {
+          this.baseNodeError(errors.toolCallTagInsideContent, node)
+        }
+
+        const attributes = await this.listTagAttributes({
+          tagNode: node as ElementTag,
+          scopeContext,
+        })
+
+        if (!attributes.has('id')) {
+          this.baseNodeError(errors.toolCallTagWithoutId, node)
+        }
+
+        if (!attributes.has('name')) {
+          this.baseNodeError(errors.toolCallWithoutName, node)
+        }
+
+        for await (const childNode of node.children ?? []) {
           await this.readBaseMetadata({
             node: childNode,
-            scopeContext: copyScopeContext(scopeContext),
+            scopeContext,
             isInMessageTag,
-            isInContentTag,
+            isInContentTag: true,
           })
         }
+
+        this.accumulatedToolCalls.push(node as ToolCallTag)
+        return
+      }
+
+      if (Object.values(ContentType).includes(node.name as ContentType)) {
+        if (isInContentTag) {
+          this.baseNodeError(errors.contentTagInsideContent, node)
+        }
+
+        for await (const childNode of node.children ?? []) {
+          await this.readBaseMetadata({
+            node: childNode,
+            scopeContext,
+            isInMessageTag,
+            isInContentTag: true,
+          })
+        }
+        return
       }
-    }
 
-    if (node.type === 'ElementTag') {
       if (
-        node.name === CUSTOM_MESSAGE_TAG ||
-        Object.values(MessageRole).includes(node.name as MessageRole)
+        Object.values(MessageRole).includes(node.name as MessageRole) ||
```

Comment:
you are repeating this Object.values(...).includes(...) in several places, dry a bit. This method is long to read and deals with much complexity, which is ok since you are solving a hard problem! make it easier for devs to understand what is happening by drying code where possible and giving meaningful names to methods.

Reply:
Right, I'll dry a bit and simplify things!

Reply:
should be clearer now!